apollo-router-core/src/json_ext.rs

```diff
@@ -138,9 +47,20 @@ impl ValueExt for Value {
                 }
             }
             (Value::Array(a), Value::Array(mut b)) => {
-                for ((_index, b_value), a_value) in b.drain(..).enumerate().zip(a.iter_mut()) {
+                for (b_value, a_value) in b.drain(..min(a.len(), b.len())).zip(a.iter_mut()) {
                     a_value.deep_merge(b_value);
                 }
+
+                a.extend(b.into_iter());
+            }
+            (_, Value::Null) => {}
+            (Value::Object(_), Value::Array(_)) => {
+                #[cfg(debug)]
+                panic!("trying to replace an object with an array");
+            }
+            (Value::Array(_), Value::Object(_)) => {
+                #[cfg(debug)]
+                panic!("trying to replace an object with an array");
```

Comment:
:screwdriver: Use `failfast_debug!()` instead. Otherwise the debug build will behave differently than the release build.

:screwdriver: The second log is upside down

```suggestion
            (Value::Object(_), Value::Array(_)) => {
                failfast_debug!("trying to replace an object with an array");
            }
            (Value::Array(_), Value::Object(_)) => {
                failfast_debug!("trying to replace an array with an object");
```